File: backend/test_utilities.py
"""
Test Utilities and Fixtures for Unit Tests (Phase 4, Step 4.1)
Provides reusable test helpers, mocks, and fixtures for all unit tests
"""

import numpy as np
import pytest
from unittest.mock import Mock, MagicMock
from io import BytesIO
import wave


class EmbeddingFixtures:
    """Fixtures for embedding-related tests"""
    
    @staticmethod
    def valid_embedding_192d():
        """Generate a valid 192-dimensional embedding"""
        return np.random.randn(192)
    
    @staticmethod
    def zero_embedding_192d():
        """Generate a zero 192-dimensional embedding"""
        return np.zeros(192)
    
    @staticmethod
    def ones_embedding_192d():
        """Generate an embedding of all ones"""
        return np.ones(192)
    
    @staticmethod
    def unit_vector_embedding():
        """Generate a normalized unit vector embedding"""
        embedding = np.random.randn(192)
        return embedding / np.linalg.norm(embedding)
    
    @staticmethod
    def embedding_with_nan():
        """Generate embedding with NaN values"""
        embedding = np.random.randn(192)
        embedding[50] = np.nan
        return embedding
    
    @staticmethod
    def embedding_with_infinity():
        """Generate embedding with Infinity values"""
        embedding = np.random.randn(192)
        embedding[75] = np.inf
        return embedding
    
    @staticmethod
    def similar_embeddings_pair(similarity_level=0.95):
        """Generate two similar embeddings"""
        base = np.random.randn(192)
        noise = np.random.randn(192) * (1 - similarity_level)
        return base, base + noise
    
    @staticmethod
    def batch_embeddings(count=100):
        """Generate batch of embeddings"""
        return [np.random.randn(192) for _ in range(count)]


class AudioFixtures:
    """Fixtures for audio-related tests"""
    
    @staticmethod
    def create_sine_wave_bytes(frequency=440, duration=1.0, sample_rate=16000):
        """Create synthetic sine wave audio bytes"""
        samples = int(duration * sample_rate)
        t = np.linspace(0, duration, samples, False)
        audio_data = (0.3 * np.sin(2 * np.pi * frequency * t)).astype(np.int16)
        
        wav_buffer = BytesIO()
        with wave.open(wav_buffer, 'wb') as wav_file:
            wav_file.setnchannels(1)
            wav_file.setsampwidth(2)
            wav_file.setframerate(sample_rate)
            wav_file.writeframes(audio_data.tobytes())
        
        wav_buffer.seek(0)
        return wav_buffer.getvalue()
    
    @staticmethod
    def create_white_noise_bytes(duration=1.0, sample_rate=16000):
        """Create white noise audio bytes"""
        samples = int(duration * sample_rate)
        audio_data = (0.1 * np.random.randn(samples)).astype(np.int16)
        
        wav_buffer = BytesIO()
        with wave.open(wav_buffer, 'wb') as wav_file:
            wav_file.setnchannels(1)
            wav_file.setsampwidth(2)
            wav_file.setframerate(sample_rate)
            wav_file.writeframes(audio_data.tobytes())
        
        wav_buffer.seek(0)
        return wav_buffer.getvalue()
    
    @staticmethod
    def create_silent_audio_bytes(duration=1.0, sample_rate=16000):
        """Create silent audio bytes"""
        samples = int(duration * sample_rate)
        audio_data = np.zeros(samples, dtype=np.int16)
        
        wav_buffer = BytesIO()
        with wave.open(wav_buffer, 'wb') as wav_file:
            wav_file.setnchannels(1)
            wav_file.setsampwidth(2)
            wav_file.setframerate(sample_rate)
            wav_file.writeframes(audio_data.tobytes())
        
        wav_buffer.seek(0)
        return wav_buffer.getvalue()
    
    @staticmethod
    def create_stereo_audio_bytes(duration=1.0, sample_rate=16000):
        """Create stereo audio bytes"""
        samples = int(duration * sample_rate)
        t = np.linspace(0, duration, samples, False)
        left_channel = (0.3 * np.sin(2 * np.pi * 440 * t)).astype(np.int16)
        right_channel = (0.3 * np.sin(2 * np.pi * 880 * t)).astype(np.int16)
        
        wav_buffer = BytesIO()
        with wave.open(wav_buffer, 'wb') as wav_file:
            wav_file.setnchannels(2)
            wav_file.setsampwidth(2)
            wav_file.setframerate(sample_rate)
            audio_data = np.column_stack((left_channel, right_channel))
            wav_file.writeframes(audio_data.tobytes())
        
        wav_buffer.seek(0)
        return wav_buffer.getvalue()


class MockDatabase:
    """Mock MongoDB database for testing"""
    
    def __init__(self):
        self.users = {}
    
    def save_user(self, phone_number, embedding):
        """Mock save_user function"""
        self.users[phone_number] = {
            "phone_number": phone_number,
            "embedding": embedding
        }
    
    def get_user(self, phone_number):
        """Mock user retrieval"""
        return self.users.get(phone_number)
    
    def verify_user(self, phone_number, query_embedding):
        """Mock user verification"""
        if phone_number not in self.users:
            return None
        
        stored_embedding = np.array(self.users[phone_number]["embedding"])
        query_array = np.array(query_embedding)
        
        # Simple cosine similarity
        score = float(np.dot(stored_embedding, query_array) / 
                     (np.linalg.norm(stored_embedding) * np.linalg.norm(query_array)))
        return score
    
    def clear(self):
        """Clear all data"""
        self.users.clear()


class ComparisonTestData:
    """Test data for comparison operations"""
    
    @staticmethod
    def similarity_test_cases():
        """Return test cases for similarity calculations"""
        np.random.seed(42)
        return [
            {
                'name': 'identical_vectors',
                'embedding1': np.ones(192),
                'embedding2': np.ones(192),
                'expected_range': (0.95, 1.0)
            },
            {
                'name': 'orthogonal_vectors',
                'embedding1': np.concatenate([np.ones(96), np.zeros(96)]),
                'embedding2': np.concatenate([np.zeros(96), np.ones(96)]),
                'expected_range': (0.3, 0.7)
            },
            {
                'name': 'random_vectors',
                'embedding1': np.random.randn(192),
                'embedding2': np.random.randn(192),
                'expected_range': (0.0, 1.0)
            }
        ]
    
    @staticmethod
    def batch_comparison_test_cases():
        """Return test cases for batch comparisons"""
        np.random.seed(42)
        base_embedding = np.random.randn(192)
        
        return {
            'enrolled': [
                base_embedding,
                base_embedding + 0.1 * np.random.randn(192),
                base_embedding + 0.2 * np.random.randn(192)
            ],
            'verification': base_embedding,
            'expected_matches': 1  # At least one should match
        }


class MetricsCollector:
    """Collect and analyze test metrics"""
    
    def __init__(self):
        self.metrics = {}
    
    def record_metric(self, name, value):
        """Record a metric value"""
        if name not in self.metrics:
            self.metrics[name] = []
        self.metrics[name].append(value)
    
    def get_average(self, name):
        """Get average of metric values"""
        if name not in self.metrics or not self.metrics[name]:
            return None
        return sum(self.metrics[name]) / len(self.metrics[name])
    
    def get_max(self, name):
        """Get maximum of metric values"""
        if name not in self.metrics or not self.metrics[name]:
            return None
        return max(self.metrics[name])
    
    def get_min(self, name):
        """Get minimum of metric values"""
        if name not in self.metrics or not self.metrics[name]:
            return None
        return min(self.metrics[name])
    
    def get_summary(self):
        """Get summary of all metrics"""
        summary = {}
        for name in self.metrics:
            summary[name] = {
                'count': len(self.metrics[name]),
                'average': self.get_average(name),
                'min': self.get_min(name),
                'max': self.get_max(name)
            }
        return summary


# Pytest Fixtures
@pytest.fixture
def embedding_fixtures():
    """Provide embedding fixtures"""
    return EmbeddingFixtures()


@pytest.fixture
def audio_fixtures():
    """Provide audio fixtures"""
    return AudioFixtures()


@pytest.fixture
def mock_database():
    """Provide mock database"""
    return MockDatabase()


@pytest.fixture
def metrics_collector():
    """Provide metrics collector"""
    return MetricsCollector()


@pytest.fixture
def test_data():
    """Provide comparison test data"""
    return ComparisonTestData()

File: backend/test_utilities.js
/**
 * Test Utilities and Fixtures for JavaScript Unit Tests (Phase 4, Step 4.1)
 * Provides reusable test helpers, mocks, and fixtures for all JavaScript unit tests
 */

/**
 * Generate valid 192-dimensional embeddings for testing
 */
const EmbeddingFixtures = {
    /**
     * Generate random 192-dimensional embedding in range [-1, 1]
     */
    validEmbedding192d() {
        return Array(192).fill(0).map(() => Math.random() * 2 - 1);
    },

    /**
     * Generate all-zeros embedding
     */
    zeroEmbedding192d() {
        return Array(192).fill(0);
    },

    /**
     * Generate all-ones embedding
     */
    onesEmbedding192d() {
        return Array(192).fill(1);
    },

    /**
     * Generate normalized unit vector embedding
     */
    unitVectorEmbedding() {
        const embedding = Array(192).fill(0).map(() => Math.random());
        const norm = Math.sqrt(
            embedding.reduce((sum, val) => sum + val * val, 0)
        );
        return embedding.map(val => val / norm);
    },

    /**
     * Generate embedding with NaN values
     */
    embeddingWithNaN() {
        const embedding = Array(192).fill(0).map(() => Math.random());
        embedding[50] = NaN;
        return embedding;
    },

    /**
     * Generate embedding with Infinity values
     */
    embeddingWithInfinity() {
        const embedding = Array(192).fill(0).map(() => Math.random());
        embedding[75] = Infinity;
        return embedding;
    },

    /**
     * Generate embedding with negative Infinity
     */
    embeddingWithNegativeInfinity() {
        const embedding = Array(192).fill(0).map(() => Math.random());
        embedding[100] = -Infinity;
        return embedding;
    },

    /**
     * Generate pair of similar embeddings
     */
    similarEmbeddingsPair(similarityLevel = 0.95) {
        const base = Array(192).fill(0).map(() => Math.random());
        const noise = Array(192).fill(0).map(() => Math.random() * (1 - similarityLevel));
        const second = base.map((val, i) => val + noise[i]);
        return [base, second];
    },

    /**
     * Generate batch of embeddings
     */
    batchEmbeddings(count = 100) {
        return Array(count).fill(0).map(() => this.validEmbedding192d());
    },

    /**
     * Generate identical embeddings for match testing
     */
    identicalEmbeddingsPair() {
        const embedding = this.validEmbedding192d();
        return [embedding, [...embedding]];
    },

    /**
     * Generate orthogonal embeddings
     */
    orthogonalEmbeddingsPair() {
        const embedding1 = Array(96).fill(1).concat(Array(96).fill(0));
        const embedding2 = Array(96).fill(0).concat(Array(96).fill(1));
        return [embedding1, embedding2];
    },

    /**
     * Generate opposite embeddings
     */
    oppositeEmbeddingsPair() {
        const embedding1 = Array(192).fill(1);
        const embedding2 = Array(192).fill(-1);
        return [embedding1, embedding2];
    },

    /**
     * Generate wrong dimension embedding
     */
    wrongDimensionEmbedding(dimension = 96) {
        return Array(dimension).fill(0.5);
    }
};

/**
 * Mock objects for testing
 */
const Mocks = {
    /**
     * Create mock logger for testing
     */
    createMockLogger() {
        return {
            info: jest.fn(),
            error: jest.fn(),
            warn: jest.fn(),
            debug: jest.fn(),
            log: jest.fn()
        };
    },

    /**
     * Create mock MongoDB collection
     */
    createMockCollection() {
        const data = {};
        return {
            insertOne: jest.fn((doc) => {
                data[doc._id] = doc;
                return Promise.resolve({ insertedId: doc._id });
            }),
            findOne: jest.fn((query) => {
                return Promise.resolve(null);
            }),
            updateOne: jest.fn((filter, update) => {
                return Promise.resolve({ modifiedCount: 1 });
            }),
            deleteOne: jest.fn((query) => {
                return Promise.resolve({ deletedCount: 1 });
            }),
            find: jest.fn(() => ({
                toArray: jest.fn(() => Promise.resolve([]))
            })),
            _data: data
        };
    },

    /**
     * Create mock Express app
     */
    createMockExpressApp() {
        return {
            get: jest.fn(),
            post: jest.fn(),
            put: jest.fn(),
            delete: jest.fn(),
            use: jest.fn(),
            listen: jest.fn((port, callback) => {
                if (callback) callback();
            })
        };
    },

    /**
     * Create mock WebSocket
     */
    createMockWebSocket() {
        return {
            on: jest.fn(),
            emit: jest.fn(),
            send: jest.fn(),
            close: jest.fn(),
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            readyState: 1,
            OPEN: 1,
            CLOSED: 3
        };
    }
};

/**
 * Test data generators
 */
const TestData = {
    /**
     * Generate similarity test cases
     */
    similarityTestCases() {
        return [
            {
                name: 'identical_vectors',
                embedding1: Array(192).fill(0.5),
                embedding2: Array(192).fill(0.5),
                expectedRange: [0.95, 1.0]
            },
            {
                name: 'orthogonal_vectors',
                embedding1: Array(96).fill(1).concat(Array(96).fill(0)),
                embedding2: Array(96).fill(0).concat(Array(96).fill(1)),
                expectedRange: [0.3, 0.7]
            },
            {
                name: 'zero_similarity',
                embedding1: Array(96).fill(1).concat(Array(96).fill(0)),
                embedding2: Array(96).fill(0).concat(Array(96).fill(1)),
                expectedRange: [0, 0.5]
            }
        ];
    },

    /**
     * Generate batch comparison test data
     */
    batchComparisonTestData() {
        const baseEmbedding = EmbeddingFixtures.validEmbedding192d();
        return {
            enrolled: [
                baseEmbedding,
                baseEmbedding.map(x => x + 0.01),
                baseEmbedding.map(x => x + 0.02)
            ],
            verification: baseEmbedding,
            expectedMatches: 1
        };
    },

    /**
     * Generate threshold test cases
     */
    thresholdTestCases() {
        return [
            { value: 0.5, isValid: true },
            { value: 0.75, isValid: true },
            { value: 1.0, isValid: true },
            { value: 0.3, isValid: false },
            { value: 1.1, isValid: false }
        ];
    }
};

/**
 * Test utilities and helpers
 */
const TestUtils = {
    /**
     * Wait for async operation
     */
    delay(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    },

    /**
     * Generate random value in range
     */
    randomInRange(min, max) {
        return Math.random() * (max - min) + min;
    },

    /**
     * Check if value is within range
     */
    isInRange(value, min, max) {
        return value >= min && value <= max;
    },

    /**
     * Compare floating point numbers
     */
    approxEqual(a, b, tolerance = 0.0001) {
        return Math.abs(a - b) < tolerance;
    },

    /**
     * Create test suite context
     */
    createTestContext() {
        return {
            startTime: Date.now(),
            metrics: {},
            recordMetric(name, value) {
                if (!this.metrics[name]) {
                    this.metrics[name] = [];
                }
                this.metrics[name].push(value);
            },
            getMetricAverage(name) {
                if (!this.metrics[name] || this.metrics[name].length === 0) {
                    return null;
                }
                return this.metrics[name].reduce((a, b) => a + b, 0) / this.metrics[name].length;
            },
            getElapsedTime() {
                return Date.now() - this.startTime;
            }
        };
    }
};

/**
 * Performance testing utilities
 */
const PerformanceUtils = {
    /**
     * Measure execution time
     */
    async measureTime(fn, label = 'Operation') {
        const startTime = Date.now();
        const result = await fn();
        const elapsed = Date.now() - startTime;
        console.log(`${label}: ${elapsed}ms`);
        return { result, elapsed };
    },

    /**
     * Benchmark function multiple times
     */
    async benchmark(fn, iterations = 100, label = 'Benchmark') {
        const times = [];
        for (let i = 0; i < iterations; i++) {
            const { elapsed } = await this.measureTime(fn, `${label} iteration ${i + 1}`);
            times.push(elapsed);
        }
        
        return {
            iterations,
            min: Math.min(...times),
            max: Math.max(...times),
            avg: times.reduce((a, b) => a + b, 0) / times.length,
            total: times.reduce((a, b) => a + b, 0),
            times
        };
    }
};

/**
 * Assertion helpers
 */
const Assertions = {
    /**
     * Assert embedding is valid 192D
     */
    assertValid192dEmbedding(embedding) {
        expect(Array.isArray(embedding)).toBe(true);
        expect(embedding.length).toBe(192);
        expect(embedding.every(val => Number.isFinite(val))).toBe(true);
    },

    /**
     * Assert similarity in valid range
     */
    assertValidSimilarity(similarity) {
        expect(typeof similarity).toBe('number');
        expect(similarity).toBeGreaterThanOrEqual(0);
        expect(similarity).toBeLessThanOrEqual(1);
    },

    /**
     * Assert result object structure
     */
    assertResultStructure(result, expectedProperties) {
        expectedProperties.forEach(prop => {
            expect(result).toHaveProperty(prop);
        });
    }
};

module.exports = {
    EmbeddingFixtures,
    Mocks,
    TestData,
    TestUtils,
    PerformanceUtils,
    Assertions,
    // Helper functions
    createTestEmbedding: () => EmbeddingFixtures.validEmbedding192d(),
    createMockLogger: () => Mocks.createMockLogger(),
    createTestContext: () => TestUtils.createTestContext()
};
